### PR TITLE
gui: add lle modules in setting dialog

### PR DESCRIPTION
**Pull request by Zangetsu38**
- Add in settings dialog Core tab with lle module for select what you want in liste.
- Add reset button for reset all modules selected.
- Add button for apply log level change without reboot.
- Add search bar for search particular module wanted on list.
- Add check for show lle list only if found one llle file.
- Fix size list language for not set item outiside setting box
# For Exemple:
image

### CODE REVIEW

**Comment by VelocityRa** [comment]
Ideally we'd do this as a multi-selectable list like RPCS3, but you probably can't do it so this is fine for now.

Other than that the English needs to be fixed and a lot of redundant text removed (I'll leave more reviews to someone else).

Also you accidentally (?) included some logging changes, remove those.

**Comment by HolyMcDiver** [comment]
I'll do the grammar checks.

File: src/emulator/gui/src/settings_dialog.cpp [hunk 1]
```diff
@@ -30,7 +30,43 @@ void draw_settings_dialog(HostState &host) {
     ImGuiTabBarFlags settings_tab_flags = ImGuiTabBarFlags_None;
     ImGui::BeginTabBar("SettingsTabBar", settings_tab_flags);
 
+    // Core
+    if (ImGui::BeginTabItem("Core")) {
+        ImGui::PopStyleColor();
+        ImGui::Text("LLE Module \nSet your LLE Module Wanted.");
```
Comment (HolyMcDiver) [inline src/emulator/gui/src/settings_dialog.cpp]: ImGui::Text("LLE Modules \nSet the desired modules for LLE.");

File: src/emulator/gui/src/settings_dialog.cpp [hunk 2]
```diff
@@ -30,7 +30,43 @@ void draw_settings_dialog(HostState &host) {
...
+        if (ImGui::IsItemHovered())
+            ImGui::SetTooltip("Write exemple 'libpgf', press Apply and re write other.");
+        std::string lle_module;
+        ImGui::InputTextWithHint("Set lle module", "Add lle module wanted", &lle_module);
```
Comment (HolyMcDiver) [inline src/emulator/gui/src/settings_dialog.cpp]: ImGui::InputTextWithHint("Set LEE module.", "Add the LLE module desired.", &lle_module);

File: src/emulator/gui/src/settings_dialog.cpp [hunk 3]
```diff
@@ -30,7 +30,43 @@ void draw_settings_dialog(HostState &host) {
...
+            if (!lle_module.empty()) {
+                host.cfg.lle_modules.push_back(lle_module);
+            }
+        }
+        ImGui::SameLine();
+        if (ImGui::Button("Reset LLE Modules")) {
+            host.cfg.lle_modules.clear();
+        }
+        static int current = -1;
+        ImGui::ListBox("LLE Module Enable", &current,
```
Comment (HolyMcDiver) [inline src/emulator/gui/src/settings_dialog.cpp]: ImGui::ListBox("LLE Modules Enabled.", &current,

File: src/emulator/gui/src/settings_dialog.cpp [hunk 4]
```diff
@@ -30,7 +30,43 @@ void draw_settings_dialog(HostState &host) {
...
+            ImGui::SetTooltip("Write exemple 'libpgf', press Apply and re write other.");
+        std::string lle_module;
+        ImGui::InputTextWithHint("Set lle module", "Add lle module wanted", &lle_module);
+        if (ImGui::Button("Apply add LLE Modules")) {
+            if (!lle_module.empty()) {
+                host.cfg.lle_modules.push_back(lle_module);
+            }
+        }
+        ImGui::SameLine();
+        if (ImGui::Button("Reset LLE Modules")) {
```
Comment (HolyMcDiver) [inline src/emulator/gui/src/settings_dialog.cpp]: if (ImGui::Button("Reset LLE Modules.")) {

File: src/emulator/gui/src/settings_dialog.cpp [hunk 5]
```diff
@@ -30,7 +30,43 @@ void draw_settings_dialog(HostState &host) {
...
 
+    // Core
+    if (ImGui::BeginTabItem("Core")) {
+        ImGui::PopStyleColor();
+        ImGui::Text("LLE Module \nSet your LLE Module Wanted.");
+        if (ImGui::IsItemHovered())
+            ImGui::SetTooltip("Write exemple 'libpgf', press Apply and re write other.");
+        std::string lle_module;
+        ImGui::InputTextWithHint("Set lle module", "Add lle module wanted", &lle_module);
+        if (ImGui::Button("Apply add LLE Modules")) {
```
Comment (HolyMcDiver) [inline src/emulator/gui/src/settings_dialog.cpp]: if (ImGui::Button("Apply LLE Modules.")) {

File: src/emulator/gui/src/settings_dialog.cpp [hunk 6]
```diff
@@ -30,7 +30,43 @@ void draw_settings_dialog(HostState &host) {
     ImGuiTabBarFlags settings_tab_flags = ImGuiTabBarFlags_None;
     ImGui::BeginTabBar("SettingsTabBar", settings_tab_flags);
 
+    // Core
+    if (ImGui::BeginTabItem("Core")) {
+        ImGui::PopStyleColor();
+        ImGui::Text("LLE Module \nSet your LLE Module Wanted.");
+        if (ImGui::IsItemHovered())
+            ImGui::SetTooltip("Write exemple 'libpgf', press Apply and re write other.");
```
Comment (HolyMcDiver) [inline src/emulator/gui/src/settings_dialog.cpp]: ImGui::SetTooltip("For example, libpgf is a LLE module. To activate any module, select it and press Apply.");

Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: Wrong, is not select for write, need write manually lle name

Comment (1whatleytay) [reply src/emulator/gui/src/settings_dialog.cpp]: So `ImGui::SetTooltip("For example, libpgf is a LLE module. To activate any module, write it and press Apply.");`?

Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: Hmm yes ^^

**Comment by Zangetsu38** [comment]
@VelocityRa change for log is normal, i have explain it on description, i have see trace is wrong and not needed for user, and so est un in info is good and can skip spam log

File: src/emulator/gui/src/settings_dialog.cpp (hunk 5 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: Is wrong, is not apply lle module is add news lle module on list ^^

File: src/emulator/gui/src/settings_dialog.cpp (hunk 4 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: I have Nevers est.  Inside butin is wrong

File: src/emulator/gui/src/settings_dialog.cpp (hunk 2 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: LEE ? 🤔

File: src/emulator/gui/src/settings_dialog.cpp (hunk 5 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: Little wrong, here is for Apply add lle inside list enable, so here said this is wrong

**Comment by Zangetsu38** [comment]
issue fixed, all works perfectly now :)

**Comment by Zangetsu38** [comment]
Code added for no can set few time same lle in list enable

File: src/emulator/gui/src/settings_dialog.cpp [hunk 7]
```diff
@@ -54,7 +54,52 @@ void draw_settings_dialog(HostState &host) {
...
     ImGui::BeginTabBar("SettingsTabBar", settings_tab_flags);
 
+    // Core
+    if (ImGui::BeginTabItem("Core")) {
+        ImGui::PopStyleColor();
+        ImGui::Text("LLE Modules \nSet the desired modules for LLE.");
+        if (ImGui::IsItemHovered())
+            ImGui::SetTooltip("For example, 'libpgf' is a LLE module.\nTo activate any module, write it and press Apply.");
+        ImGui::InputTextWithHint("Set LLE Module", "Add the LLE module desired.", &host.gui.lle_module);
+        if (ImGui::Button("Apply Adding LLE Module")) {
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
        if (ImGui::Button("Add")) {
```

File: src/emulator/gui/src/settings_dialog.cpp [hunk 8]
```diff
@@ -54,7 +54,52 @@ void draw_settings_dialog(HostState &host) {
     ImGuiTabBarFlags settings_tab_flags = ImGuiTabBarFlags_None;
     ImGui::BeginTabBar("SettingsTabBar", settings_tab_flags);
 
+    // Core
+    if (ImGui::BeginTabItem("Core")) {
+        ImGui::PopStyleColor();
+        ImGui::Text("LLE Modules \nSet the desired modules for LLE.");
+        if (ImGui::IsItemHovered())
+            ImGui::SetTooltip("For example, 'libpgf' is a LLE module.\nTo activate any module, write it and press Apply.");
+        ImGui::InputTextWithHint("Set LLE Module", "Add the LLE module desired.", &host.gui.lle_module);
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
        ImGui::InputTextWithHint("", "Module name.", &host.gui.lle_module);
```

File: src/emulator/gui/src/settings_dialog.cpp [hunk 9]
```diff
@@ -54,7 +54,52 @@ void draw_settings_dialog(HostState &host) {
     ImGuiTabBarFlags settings_tab_flags = ImGuiTabBarFlags_None;
     ImGui::BeginTabBar("SettingsTabBar", settings_tab_flags);
 
+    // Core
+    if (ImGui::BeginTabItem("Core")) {
+        ImGui::PopStyleColor();
+        ImGui::Text("LLE Modules \nSet the desired modules for LLE.");
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
        ImGui::Text("LLE Modules \nSet the desired modules to be loaded.");
```

File: src/emulator/gui/include/gui/state.h [hunk 10]
```diff
@@ -93,6 +93,7 @@ struct State {
     char disassembly_count[5] = "100";
     std::vector<std::string> disassembly;
 
+    std::string lle_module;
```
Comment (VelocityRa) [inline src/emulator/gui/include/gui/state.h]: Why is this here? Just make it a local in `draw_settings_dialog`.

File: src/emulator/config.cpp [hunk 11]
```diff
@@ -163,7 +163,7 @@ static ExitCode parse(Config &cfg, fs::path load_path, const std::string &root_p
     get_yaml_value(config_node, "sys-lang", &cfg.sys_lang, static_cast<int>(SCE_SYSTEM_PARAM_LANG_ENGLISH_US));
     get_yaml_value(config_node, "background-image", &cfg.background_image, std::string{});
     get_yaml_value(config_node, "background-alpha", &cfg.background_alpha, 0.300f);
-    get_yaml_value(config_node, "log-level", &cfg.log_level, static_cast<int>(spdlog::level::trace));
+    get_yaml_value(config_node, "log-level", &cfg.log_level, static_cast<int>(spdlog::level::info));
```
Comment (VelocityRa) [inline src/emulator/config.cpp]: remove

File: src/emulator/host/include/host/config.h [hunk 12]
```diff
@@ -31,7 +31,7 @@ struct Config {
     optional<std::string> vpk_path;
     optional<std::string> run_title_id;
     optional<std::string> recompile_shader_path;
-    int log_level = 0;
+    int log_level = 2;
```
Comment (VelocityRa) [inline src/emulator/host/include/host/config.h]: remove

File: src/emulator/gui/src/settings_dialog.cpp [hunk 13]
```diff
@@ -87,6 +132,10 @@ void draw_settings_dialog(HostState &host) {
         ImGui::Combo("Log Level", &host.cfg.log_level, "Trace\0Debug\0Info\0Warning\0Error\0Critical\0Off\0");
         if (ImGui::IsItemHovered())
             ImGui::SetTooltip("Select your preferred log level.");
+        if (ImGui::Button("Apply change")) {
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]: No need for this, just apply every time a module is added/deleted.

File: src/emulator/gui/src/settings_dialog.cpp [hunk 14]
```diff
@@ -54,7 +54,52 @@ void draw_settings_dialog(HostState &host) {
...
+                [](void *vec, int idx, const char **out_text) {
+                    auto &vector = *static_cast<std::vector<std::string> *>(vec);
+                    if (idx < 0 || idx >= static_cast<int>(vector.size())) {
+                        return false;
+                    }
+                    *out_text = vector.at(idx).c_str();
+                    return true;
+                },
+                static_cast<void *>(&host.cfg.lle_modules), host.cfg.lle_modules.size());
+            if (ImGui::Button("Reset LLE Modules")) {
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
            if (ImGui::Button("Clear list")) {
```

File: src/emulator/gui/src/settings_dialog.cpp [hunk 15]
```diff
@@ -54,7 +54,52 @@ void draw_settings_dialog(HostState &host) {
...
+                host.cfg.lle_modules.clear();
+            }
+            ImGui::SameLine();
+            if (ImGui::Button("Remove last LLE Module")) {
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
            if (ImGui::Button("Remove last entry")) {
```

File: src/emulator/config.cpp (hunk 11 above)
Comment (Zangetsu38) [reply src/emulator/config.cpp]: why remove ? i have set it for normal user for get info byu default, like rpcs3 on succes, can change it after, is just default value for normal user

Comment (VelocityRa) [reply src/emulator/config.cpp]: No. We want as much info as possible.
Even it was an acceptable change, why even include it in the same commit???

File: src/emulator/gui/include/gui/state.h (hunk 10 above)
Comment (Zangetsu38) [reply src/emulator/gui/include/gui/state.h]: Is @1whatleytay have said me need set here, for get constant value and no lost input set

File: src/emulator/config.cpp (hunk 11 above)
Comment (Zangetsu38) [reply src/emulator/config.cpp]: ah yes i see probleme artf, i have thinks works like rpcs3, but here set info is not oinly for showing log, but only for print inside log file arf
i have removed

File: src/emulator/gui/src/settings_dialog.cpp (hunk 13 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: no, here is for change log level without reboot no relate lle module

File: src/emulator/gui/src/settings_dialog.cpp [hunk 16]
```diff
@@ -54,7 +54,52 @@ void draw_settings_dialog(HostState &host) {
...
+                    host.cfg.lle_modules.push_back(host.gui.lle_module);
+                    host.gui.lle_module.clear();
+                } else
+                    LOG_WARN("Module already set");
+            }
+        }
+        ImGui::Spacing();
+        if (!host.cfg.lle_modules.empty()) {
+            static int current = -1;
+            ImGui::ListBox("List Modules Enabled", &current,
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
            ImGui::ListBox("##module_list", &current,
```

File: src/emulator/gui/include/gui/state.h (hunk 10 above)
Comment (Zangetsu38) [reply src/emulator/gui/include/gui/state.h]: i have founded how get constant value inside settings dialog

**Comment by Zangetsu38** [comment]
Change pushed and  screen example updated

**Comment by Zangetsu38** [comment]
i push soon commit for reworks all that and using directly list with module name for preview
image

**Comment by Zangetsu38** [comment]
Pushed finally news version with using list multi selectable :)
news screenshot for exemple on description

File: src/emulator/gui/src/settings_dialog.cpp [hunk 17]
```diff
@@ -48,13 +49,83 @@ bool change_pref_location(const std::string &input_path, const std::string &curr
...
     ImGui::PushStyleColor(ImGuiCol_Text, GUI_COLOR_TEXT_MENUBAR_OPTIONS);
     ImGui::Begin("Settings", &host.gui.configuration_menu.settings_dialog, ImGuiWindowFlags_AlwaysAutoResize);
     ImGuiTabBarFlags settings_tab_flags = ImGuiTabBarFlags_None;
     ImGui::BeginTabBar("SettingsTabBar", settings_tab_flags);
 
+    // Core
+    if (ImGui::BeginTabItem("Core")) {
+        ImGui::PopStyleColor();
+        if (vfs::read_file(VitaIoDevice::VS0, buffer, host.pref_path, "sys/external/libc.suprx")) {
+            ImGui::Text("List Modules \nSelect your modules desired.");
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
            ImGui::Text("Module List\nSelect your desired modules.");
```

File: src/emulator/gui/src/settings_dialog.cpp [hunk 18]
```diff
@@ -48,13 +49,83 @@ bool change_pref_location(const std::string &input_path, const std::string &curr
...
+            static const char *list_modules[] = {
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
            static const char *MODULE_LIST[] = {
```

File: src/emulator/gui/src/settings_dialog.cpp [hunk 19]
```diff
@@ -48,13 +49,83 @@ bool change_pref_location(const std::string &input_path, const std::string &curr
...
+
 void draw_settings_dialog(HostState &host) {
     ImGui::PushStyleColor(ImGuiCol_Text, GUI_COLOR_TEXT_MENUBAR_OPTIONS);
     ImGui::Begin("Settings", &host.gui.configuration_menu.settings_dialog, ImGuiWindowFlags_AlwaysAutoResize);
     ImGuiTabBarFlags settings_tab_flags = ImGuiTabBarFlags_None;
     ImGui::BeginTabBar("SettingsTabBar", settings_tab_flags);
 
+    // Core
+    if (ImGui::BeginTabItem("Core")) {
+        ImGui::PopStyleColor();
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]: remove, push the color you need if you need another one, and pop at the end of the `if`.

File: src/emulator/gui/src/settings_dialog.cpp [hunk 20]
```diff
@@ -48,13 +49,83 @@ bool change_pref_location(const std::string &input_path, const std::string &curr
...
+            ImGui::PopItemWidth();
+            ImGui::Spacing();
+            ImGui::TextColored(GUI_COLOR_TEXT, "Search Modules");
+            search_bar.Draw("##search_bar", 200);
+            ImGui::Spacing();
+            if (ImGui::Button("Clear list"))
+                host.cfg.lle_modules.clear();
+        }
+        ImGui::EndTabItem();
+    } else {
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]: remove this

File: src/emulator/gui/src/settings_dialog.cpp [hunk 21]
```diff
@@ -48,13 +49,83 @@ bool change_pref_location(const std::string &input_path, const std::string &curr
...
+        ImGui::PopStyleColor();
+    }
+
     // System
+    ImGui::PushStyleColor(ImGuiCol_Text, GUI_COLOR_TEXT_MENUBAR_OPTIONS);
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]: remove

File: src/emulator/gui/src/settings_dialog.cpp [hunk 22]
```diff
@@ -87,6 +158,10 @@ void draw_settings_dialog(HostState &host) {
         ImGui::Combo("Log Level", &host.cfg.log_level, "Trace\0Debug\0Info\0Warning\0Error\0Critical\0Off\0");
         if (ImGui::IsItemHovered())
             ImGui::SetTooltip("Select your preferred log level.");
+        if (ImGui::Button("Apply change")) {
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
        if (ImGui::Button("Apply")) {
```

File: src/emulator/gui/src/settings_dialog.cpp (hunk 18 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: Why in shift ? list_sys_lang is not shift

File: src/emulator/gui/src/settings_dialog.cpp (hunk 19 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: Is not possible @pent0 have help me first time usine like that is obligarory

File: src/emulator/gui/src/settings_dialog.cpp (hunk 21 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: Same need add push coloré before all tab

File: src/emulator/gui/src/settings_dialog.cpp (hunk 20 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: remove style for thsi tab broken all i can't

File: src/emulator/gui/src/settings_dialog.cpp [hunk 23]
```diff
@@ -48,13 +49,83 @@ bool change_pref_location(const std::string &input_path, const std::string &curr
...
     ImGui::BeginTabBar("SettingsTabBar", settings_tab_flags);
 
+    // Core
+    if (ImGui::BeginTabItem("Core")) {
+        ImGui::PopStyleColor();
+        if (vfs::read_file(VitaIoDevice::VS0, buffer, host.pref_path, "sys/external/libc.suprx")) {
+            ImGui::Text("Modules Listes\nSelect your desired modules.");
+            static const char *list_modules[] = {
+                // clang-format off
+                "activity_db", "adhoc_matching", "apputil", "apputil_ext", "audiocodec", "avcdec_for_player", "bgapputil", "bXCe",
```
Comment (pent0) [inline src/emulator/gui/src/settings_dialog.cpp]: Put this into a "inc" file or something. I can walk you through on Discord later

File: src/emulator/gui/src/settings_dialog.cpp (hunk 18 above)
Comment (pent0) [reply src/emulator/gui/src/settings_dialog.cpp]: const things should always be capitalized

Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: all list inside imgui demo don't have capitalized, and list_sys_lang is like this also, and whith my search on web for code some thing, i have see very lot list like this, and never see shifted

File: src/emulator/gui/src/settings_dialog.cpp (hunk 23 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: yes, i have thinks same but no have idea how set that hmm

File: src/emulator/gui/src/settings_dialog.cpp (hunk 18 above)
Comment (pent0) [reply src/emulator/gui/src/settings_dialog.cpp]: We are unique. Also this is static const, like constexpr. If there are places where it's static const but isnt capitalized, we can change it

File: src/emulator/gui/src/settings_dialog.cpp [hunk 24]
```diff
@@ -48,13 +49,83 @@ bool change_pref_location(const std::string &input_path, const std::string &curr
     return true;
 }
 
+std::vector<std::string>::iterator modules;
+vfs::FileBuffer buffer;
+static ImGuiTextFilter search_bar;
```
Comment (pent0) [inline src/emulator/gui/src/settings_dialog.cpp]: Bold reason to put this global. Please avoid doing this when you can.
Try to put these variables to function

File: src/emulator/gui/src/settings_dialog.cpp (hunk 19 above)
Comment (pent0) [reply src/emulator/gui/src/settings_dialog.cpp]: The push/pop is suppose for the color of the label of the tab button. I think this is valid

File: src/emulator/gui/src/settings_dialog.cpp [hunk 25]
```diff
@@ -48,13 +49,83 @@ bool change_pref_location(const std::string &input_path, const std::string &curr
...
+                                host.cfg.lle_modules.erase(modules);
+                        }
+                    }
+                }
+                ImGui::ListBoxFooter();
+            }
+            ImGui::PopItemWidth();
+            ImGui::Spacing();
+            ImGui::TextColored(GUI_COLOR_TEXT, "Search Modules");
+            search_bar.Draw("##search_bar", 200);
```
Comment (pent0) [inline src/emulator/gui/src/settings_dialog.cpp]: Again, i know you are putting this global in order to reuse, put the search bar variable are created very quick. Just put another one inside here

**Comment by Zangetsu38** [comment]
I have pushed very big change, with refactoring code and set lot things inside news header, for me is very more proper like that.
Need review please

File: src/emulator/gui/include/gui/settings_dialog.h [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+#include <host/state.h>
+#include <io/io.h>
+#include <misc/cpp/imgui_stdlib.h>
+#include <util/fs.h>
+#include <util/log.h>
+
+namespace gui {
+vfs::FileBuffer buffer;
+static ImGuiTextFilter search_bar;
+std::vector<std::string>::iterator modules;
```
Comment (pent0) [inline src/emulator/gui/include/gui/settings_dialog.h]: Is this a joke?

File: src/emulator/gui/include/gui/settings_dialog.h [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+            boost::system::error_code err;
+            it.increment(err);
+        }
+    } catch (const std::exception &err) {
+        return false;
+    }
+    return true;
+}
+
+static const char *LIST_MODULES[] = {
```
Comment (pent0) [inline src/emulator/gui/include/gui/settings_dialog.h]: I guess this is acceptable then. I will refeactor it one day.

File: src/emulator/gui/src/settings_dialog.cpp [hunk 28]
```diff
@@ -15,53 +15,61 @@
...
+            ImGui::PushItemWidth(240);
+            if (ImGui::ListBoxHeader("##modules_list", IM_ARRAYSIZE(LIST_MODULES), 8)) {
+                static bool module_select[IM_ARRAYSIZE(LIST_MODULES)] = { false };
+                for (int m = 0; m < IM_ARRAYSIZE(LIST_MODULES); m++) {
+                    if (!search_bar.PassFilter(LIST_MODULES[m]))
+                        continue;
+                    modules = std::find(host.cfg.lle_modules.begin(), host.cfg.lle_modules.end(), LIST_MODULES[m]);
+                    if (modules != host.cfg.lle_modules.end())
+                        module_select[m] = true;
+                    else if (modules == host.cfg.lle_modules.end())
```
Comment (pent0) [inline src/emulator/gui/src/settings_dialog.cpp]: Unneccessary else

File: src/emulator/gui/src/settings_dialog.cpp [hunk 29]
```diff
@@ -15,53 +15,61 @@
...
     ImGuiTabBarFlags settings_tab_flags = ImGuiTabBarFlags_None;
     ImGui::BeginTabBar("SettingsTabBar", settings_tab_flags);
 
+    // Core
+    if (ImGui::BeginTabItem("Core")) {
+        ImGui::PopStyleColor();
+        if (vfs::read_file(VitaIoDevice::VS0, buffer, host.pref_path, "sys/external/libc.suprx")) {
+            ImGui::Text("Modules List\nSelect your desired modules.");
+            ImGui::PushItemWidth(240);
+            if (ImGui::ListBoxHeader("##modules_list", IM_ARRAYSIZE(LIST_MODULES), 8)) {
```
Comment (pent0) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
            constexpr int MODULES_COUNT = IM_ARRAYSIZE(LIST_MODULES);
            if (ImGui::ListBoxHeader("##modules_list", MODULES_COUNT, 8)) {
```

Since you are reusing that phrase

Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: Why remove listboxheader ? I have set it for get it works like recommanded imgui creator for using with selectable

Comment (pent0) [reply src/emulator/gui/src/settings_dialog.cpp]: I dont tell you to remove it. Look carefully.

File: src/emulator/gui/src/settings_dialog.cpp (hunk 28 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: You said this complet line or just else Alone ?

File: src/emulator/gui/include/gui/settings_dialog.h (hunk 26 above)
Comment (Zangetsu38) [reply src/emulator/gui/include/gui/settings_dialog.h]: Why joke ? I have create this header for that for move this here for can clean cpp file

Comment (pent0) [reply src/emulator/gui/include/gui/settings_dialog.h]: I told you to move it into a function.

File: src/emulator/gui/src/settings_dialog.cpp (hunk 28 above)
Comment (pent0) [reply src/emulator/gui/src/settings_dialog.cpp]: the complete else block

File: src/emulator/gui/include/gui/settings_dialog.h (hunk 26 above)
Comment (Zangetsu38) [reply src/emulator/gui/include/gui/settings_dialog.h]: yes i know, but like i said, now i have create header, so set this in function is not needed.
Is same for move part change_pref_location here, can clean up cpp file with not needed code inside

File: src/emulator/gui/src/settings_dialog.cpp (hunk 28 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: ah, i have set all that because teste have show me is needed, but, i have thinks i can move this in other part

File: src/emulator/gui/include/gui/settings_dialog.h (hunk 26 above)
Comment (pent0) [reply src/emulator/gui/include/gui/settings_dialog.h]: Put it in the function.

File: src/emulator/gui/src/settings_dialog.cpp (hunk 29 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: ah yes wrong see lol sorry

File: src/emulator/gui/src/settings_dialog.cpp (hunk 28 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: after test set it other part, i can't remove it here, is obligatory for function clear list, without this use clear list not set select module in false

**Comment by Zangetsu38** [comment]
Pushed change request

File: src/emulator/gui/include/gui/settings_dialog.h (hunk 26 above)
Comment (Zangetsu38) [reply src/emulator/gui/include/gui/settings_dialog.h]: iu have set like you want, i have juust wanted set here for more "clean" code inside gui code only

File: src/emulator/gui/src/settings_dialog.cpp [hunk 30]
```diff
@@ -15,53 +15,72 @@
...
+            }
+            ImGui::PopItemWidth();
+            ImGui::Spacing();
+            ImGui::TextColored(GUI_COLOR_TEXT, "Search Modules");
+            search_bar.Draw("##search_bar", 200);
+            ImGui::Spacing();
+            if (ImGui::Button("Clear list"))
+                host.cfg.lle_modules.clear();
+        } else {
+            ImGui::Text("Modules list is Hiden.\nPlease dump decrypted modules from your PS Vita.");
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
            ImGui::Text("No modules present.\nPlease dump decrypted modules from your PS Vita (see the "#info-faq" channel in our Discord server for details)");
```

File: src/emulator/gui/src/settings_dialog.cpp [hunk 31]
```diff
@@ -15,53 +15,72 @@
...
     ImGui::PushStyleColor(ImGuiCol_Text, GUI_COLOR_TEXT_MENUBAR_OPTIONS);
     ImGui::Begin("Settings", &host.gui.configuration_menu.settings_dialog, ImGuiWindowFlags_AlwaysAutoResize);
     ImGuiTabBarFlags settings_tab_flags = ImGuiTabBarFlags_None;
     ImGui::BeginTabBar("SettingsTabBar", settings_tab_flags);
 
+    // Core
+    if (ImGui::BeginTabItem("Core")) {
+        ImGui::PopStyleColor();
+        if (vfs::read_file(VitaIoDevice::VS0, buffer, host.pref_path, "sys/external/libc.suprx")) {
+            ImGui::Text("Modules List");
```
Comment (VelocityRa) [inline src/emulator/gui/src/settings_dialog.cpp]:
```suggestion
            ImGui::Text("Module list");
```

File: src/emulator/gui/include/gui/settings_dialog.h [hunk 32]
```diff
@@ -0,0 +1,82 @@
...
+#include <host/functions.h>
+#include <host/state.h>
+#include <io/io.h>
+#include <misc/cpp/imgui_stdlib.h>
+#include <util/fs.h>
+#include <util/log.h>
+
+namespace gui {
+
+bool change_pref_location(const std::string &input_path, const std::string &current_path) {
```
Comment (VelocityRa) [inline src/emulator/gui/include/gui/settings_dialog.h]: If you're not gonna implement this, just remove this function.

Comment (Zangetsu38) [reply src/emulator/gui/include/gui/settings_dialog.h]: is not me that, i have just move code inside header for clean cpp

File: src/emulator/gui/src/settings_dialog.cpp (hunk 31 above)
Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: hmm Use Module is wrong no ? because is several modules so have S

Comment (VelocityRa) [reply src/emulator/gui/src/settings_dialog.cpp]: In English it be either, and without an "s" is more common.

File: src/emulator/gui/include/gui/settings_dialog.h (hunk 32 above)
Comment (Zangetsu38) [reply src/emulator/gui/include/gui/settings_dialog.h]: and this function is used inside cpp file for copy file old folder

Comment (pent0) [reply src/emulator/gui/include/gui/settings_dialog.h]: This was called somewhere

Comment (VelocityRa) [reply src/emulator/gui/include/gui/settings_dialog.h]: Leave it as it is inside the .cpp file.
.cpp files are for implementations, not .h files.

Comment (Zangetsu38) [reply src/emulator/gui/include/gui/settings_dialog.h]: ah hmm i have juust thinks use like before is very uggly :/

Comment (Zangetsu38) [reply src/emulator/gui/include/gui/settings_dialog.h]: I have restored

**Comment by Zangetsu38** [comment]
Push change for don't using read_file but fs::exist  for found if folder module exist

File: src/emulator/gui/include/gui/settings_dialog.h [hunk 33]
```diff
@@ -0,0 +1,61 @@
...
+    "np_message", "np_message_contacts", "np_message_dialog_impl", "np_message_padding", "np_party", "np_ranking",
+    "np_signaling", "np_sns_facebook", "np_trophy", "np_tus", "np_utility", "np_webapi", "party_member_list",
+    "psmkdc", "pspnet_adhoc", "signin_ext", "sqlite", "store_checkout_plugin", "trigger_util", "web_ui_plugin"
+};
+
+static const char *LIST_SYS_LANG[] = {
+    "Japanese", "American English", "French", "Spanish", "German", "Italian", "Dutch", "Portugal Portuguese", "Russian", "Korean",
+    "Traditional Chinese", "Simplified Chinese", "Finnish", "Swedish", "Danish", "Norwegian", "Polish", "Brazil Portuguese", "British English", "Turkish"
+};
+} // namespace gui
```
Comment (petmac) [inline src/emulator/gui/include/gui/settings_dialog.h]: Looks like this file is private to the settings dialog. It could get moved out of the public header folder and into `gui/src`.

Or since it's only used from `settings_dialog.cpp` just inline it.

File: src/emulator/gui/src/settings_dialog.cpp [hunk 34]
```diff
@@ -49,19 +41,66 @@ bool change_pref_location(const std::string &input_path, const std::string &curr
 }
 
 void draw_settings_dialog(HostState &host) {
+    static ImGuiTextFilter search_bar;
```
Comment (petmac) [inline src/emulator/gui/src/settings_dialog.cpp]: This should probably be in the GUI state structure rather than static.

Comment (Zangetsu38) [reply src/emulator/gui/src/settings_dialog.cpp]: hmm i see, just other dev have said me put all adding line here

File: src/emulator/gui/include/gui/settings_dialog.h (hunk 33 above)
Comment (Zangetsu38) [reply src/emulator/gui/include/gui/settings_dialog.h]: dev no have wanted i put it inside cpp file, and for using this, i am obligatory need use static const char, so i have thinks put it inside header for setting dialogs, and get coede inside cpp very more clear like that :)

**Comment by Zangetsu38** [comment]
Pushed refactor works after reviewed by @petmac thx

**Comment by Zangetsu38** [comment]
Refactor part code selabtle and unselectable for get more small and more smart with check also in open dialog setting for not spam loop enable or disable.

**Comment by Zangetsu38** [comment]
Pushed refactor game search bar with remove not needed code, and set like modules search.
And remove else if for set selected in off with used memet clear like demo imgui, for it call only one time, more clean
